File: libraries/lend-shared/src/interfaces/market.rs
use fadroma::{
    admin, cosmwasm_std, derive_contract::*, killswitch, schemars, schemars::JsonSchema,
    snip20_impl::msg::QueryAnswer as Snip20Response, to_binary, Binary, Callback, ContractLink,
    Decimal256, HandleResponse, HumanAddr, InitResponse, Querier, QueryRequest, StdError,
    StdResult, Uint128, Uint256, WasmQuery,
};

use serde::{Deserialize, Serialize};

use crate::interfaces::overseer::{
    AccountLiquidity,
    Market as EnteredMarket
};
use crate::core::{MasterKey, AuthMethod, Pagination};

#[interface(
    component(path = "admin"),
    component(path = "killswitch")
)]
pub trait Market {
    #[init]
    fn new(
        admin: HumanAddr,
        prng_seed: Binary,
        entropy: Binary,
        key: MasterKey,
        // Underlying asset address
        underlying_asset: ContractLink<HumanAddr>,
        // Interest model contract address
        interest_model_contract: ContractLink<HumanAddr>,
        config: Config,
        callback: Callback<HumanAddr>,
    ) -> StdResult<InitResponse>;

    /// Snip20 receiver interface
    #[handle]
    fn receive(
        sender: HumanAddr,
        from: HumanAddr,
        msg: Option<Binary>,
        amount: Uint128,
    ) -> StdResult<HandleResponse>;

    #[handle]
    fn redeem_token(burn_amount: Uint256) -> StdResult<HandleResponse>;

    #[handle]
    fn redeem_underlying(receive_amount: Uint256) -> StdResult<HandleResponse>;

    #[handle]
    fn borrow(amount: Uint256) -> StdResult<HandleResponse>;

    #[handle]
    fn transfer(recipient: HumanAddr, amount: Uint256) -> StdResult<HandleResponse>;

    #[handle]
    fn send(
        recipient: HumanAddr,
        recipient_code_hash: Option<String>,
        amount: Uint256,
        msg: Option<Binary>,
        memo: Option<String>,
        padding: Option<String>,
    ) -> StdResult<HandleResponse>;

    #[handle]
    fn register_receive(
        code_hash: String, 
        padding: Option<String>
    ) -> StdResult<HandleResponse>;

    #[handle]
    fn accrue_interest() -> StdResult<HandleResponse>;

    #[handle]
    fn seize(
        liquidator: HumanAddr,
        borrower: HumanAddr,
        amount: Uint256,
    ) -> StdResult<HandleResponse>;

    #[handle]
    fn update_config(
        interest_model: Option<ContractLink<HumanAddr>>,
        reserve_factor: Option<Decimal256>,
        borrow_cap: Option<Uint256>,
    ) -> StdResult<HandleResponse>;

    #[handle]
    fn reduce_reserves(amount: Uint128, to: Option<HumanAddr>) -> StdResult<HandleResponse>;

    #[handle]
    fn create_viewing_key(entropy: String, padding: Option<String>) -> StdResult<HandleResponse>;

    #[handle]
    fn set_viewing_key(key: String, padding: Option<String>) -> StdResult<HandleResponse>;

    #[query]
    fn token_info() -> StdResult<Snip20Response>;

    #[query]
    fn balance(address: HumanAddr, key: String) -> StdResult<Uint128>;

    #[query]
    fn balance_underlying(method: MarketAuth, block: Option<u64>) -> StdResult<Uint128>;

    #[query]
    fn state(block: Option<u64>) -> StdResult<State>;

    #[query]
    fn underlying_asset() -> StdResult<ContractLink<HumanAddr>>;

    #[query]
    fn borrow_rate(block: Option<u64>) -> StdResult<Decimal256>;

    #[query]
    fn supply_rate(block: Option<u64>) -> StdResult<Decimal256>;

    #[query]
    fn exchange_rate(block: Option<u64>) -> StdResult<Decimal256>;

    #[query]
    fn account(method: MarketAuth, block: Option<u64>) -> StdResult<AccountInfo>;

    #[query]
    fn id(method: MarketAuth) -> StdResult<Binary>;

    #[query]
    fn borrowers(
        block: u64,
        pagination: Pagination
    ) -> StdResult<BorrowersResponse>;
}

pub type MarketAuth = AuthMethod<MarketPermissions>;

#[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
#[serde(rename_all = "snake_case")]
pub enum MarketPermissions {
    AccountInfo,
    Balance,
    Id,
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub struct State {
    /// Block number that the interest was last accrued at
    pub accrual_block: u64,
    /// Accumulator of the total earned interest rate since the opening of the market
    pub borrow_index: Decimal256,
    /// Total amount of outstanding borrows of the underlying in this market
    pub total_borrows: Uint256,
    /// Total amount of reserves of the underlying held in this market
    pub total_reserves: Uint256,
    /// Total number of tokens in circulation
    pub total_supply: Uint256,
    /// The amount of the underlying token that the market has.
    pub underlying_balance: Uint128,
    /// Values in the contract that rarely change.
    pub config: Config,
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct Config {
    /// Initial exchange rate used when minting the first slTokens (used when totalSupply = 0)
    pub initial_exchange_rate: Decimal256,
    /// Fraction of interest currently set aside for reserves
    pub reserve_factor: Decimal256,
    /// Share of seized collateral that is added to reserves
    pub seize_factor: Decimal256,
}

impl Config {
    pub fn set_reserve_factor(&mut self, new: Decimal256) -> StdResult<()> {
        Self::validate_reserve_factor(&new)?;

        self.reserve_factor = new;

        Ok(())
    }

    fn validate_reserve_factor(reserve_factor: &Decimal256) -> StdResult<()> {
        if *reserve_factor > Decimal256::one() {
            return Err(StdError::generic_err("Reserve factor must be lower than or equal to 1"));
        } else {
            Ok(())
        }
    }

    fn validate_initial_exchange_rate(rate: &Decimal256) -> StdResult<()> {
        if *rate == Decimal256::zero() {
            return Err(StdError::generic_err("Initial exchange rate must be greater than 0"));
        } else {
            Ok(())
        }
    }

    pub fn validate(&self) -> StdResult<()> {
        Self::validate_initial_exchange_rate(&self.initial_exchange_rate)?;
        Self::validate_reserve_factor(&self.reserve_factor)?;

        Ok(())
    }
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct BorrowersResponse {
    pub entries: Vec<Borrower>,
    /// The total number of entries stored by the contract.
    pub total: u64
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct Borrower {
    pub id: Binary,
    pub principal_balance: Uint256,
    pub actual_balance: Uint256,
    pub liquidity: AccountLiquidity,
    pub markets: Vec<EnteredMarket<HumanAddr>>,
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Default, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct BorrowerInfo {
    /// Total balance (with accrued interest), after applying the most recent balance-changing action
    pub principal: Uint256,
    /// Global borrowIndex as of the most recent balance-changing action
    pub interest_index: Decimal256,
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
#[serde(rename_all = "snake_case")]
#[serde(deny_unknown_fields)]
pub struct AccountInfo {
    pub sl_token_balance: Uint256,
    pub borrow_balance: Uint256,
    pub exchange_rate: Decimal256,
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
#[serde(rename_all = "snake_case")]
#[serde(deny_unknown_fields)]
pub enum ReceiverCallbackMsg {
    /// Deposit underlying token
    Deposit,
    Repay {
        /// Repay someone else's debt.
        borrower: Option<Binary>,
    },
    Liquidate {
        borrower: Binary,
        collateral: HumanAddr,
    },
}

pub fn query_exchange_rate(
    querier: &impl Querier,
    market: ContractLink<HumanAddr>,
    block: Option<u64>,
) -> StdResult<Decimal256> {
    querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
        contract_addr: market.address,
        callback_code_hash: market.code_hash,
        msg: to_binary(&QueryMsg::ExchangeRate { block })?,
    }))
}

pub fn query_account(
    querier: &impl Querier,
    market: ContractLink<HumanAddr>,
    method: MarketAuth,
    block: Option<u64>,
) -> StdResult<AccountInfo> {
    querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
        contract_addr: market.address,
        callback_code_hash: market.code_hash,
        msg: to_binary(&QueryMsg::Account { method, block })?,
    }))
}

File: frontends/siennajs/lib/lend/market.ts
import { SmartContract, Querier } from '../contract'
import { Snip20Contract, TokenInfo } from '../snip20'
import { AccountLiquidity, Market } from './overseer'
import {
    Fee, Address, ContractInfo, Pagination, Decimal256,
    Uint256, Uint128, ViewingKey, create_fee
} from '../core'
import { ViewingKeyExecutor } from '../executors/viewing_key_executor'
import { LendAuth } from './auth'
import { PaginatedResponse } from '.'

import { ExecuteResult, SigningCosmWasmClient } from 'secretjs'

export interface MarketState {
    /**
     * Block height that the interest was last accrued at
     */
    accrual_block: number,
    /**
     * Accumulator of the total earned interest rate since the opening of the market
     */
    borrow_index: Decimal256,
    /**
     * Total amount of outstanding borrows of the underlying in this market
     */
    total_borrows: Uint256,
    /**
     * Total amount of reserves of the underlying held in this market
     */
    total_reserves: Uint256,
    /**
     * Total number of tokens in circulation
     */
    total_supply: Uint256,
    /**
     * The amount of the underlying token that the market has.
     */
    underlying_balance: Uint128,
    /**
     * Values in the contract that rarely change.
     */
    config: MarketConfig
}

export interface MarketConfig {
    /**
     * Initial exchange rate used when minting the first slTokens (used when totalSupply = 0)
     */
    initial_exchange_rate: Decimal256,
    /**
     * Fraction of interest currently set aside for reserves
     */
    reserve_factor: Decimal256,
    /**
     * Share of seized collateral that is added to reserves
     */
    seize_factor: Decimal256
}

export interface MarketAccount {
    /**
     * Amount of slToken that this account has.
     */
    sl_token_balance: Uint256,
    /**
     * How much the account has borrowed.
     */
    borrow_balance: Uint256,
    /**
     * The current exchange rate in the market.
     */
    exchange_rate: Decimal256
}

export interface MarketBorrower {
    id: string,
    /**
     * Borrow balance at the last interaction of the borrower.
     */
    principal_balance: Uint256,
    /**
     * Current borrow balance.
     */
    actual_balance: Uint256,
    liquidity: AccountLiquidity,
    markets: Market[]
}

export type MarketPermissions = 'account_info' | 'balance' | 'id'

export class MarketContract extends SmartContract<MarketExecutor, MarketQuerier> {
    exec(fee?: Fee, memo?: string): MarketExecutor {
        return new MarketExecutor(
            this.address,
            () => this.query.apply(this),
            this.execute_client,
            fee,
            memo
        )
    }

    query(): MarketQuerier {
        return new MarketQuerier(this.address, this.query_client)
    }
}

class MarketExecutor extends ViewingKeyExecutor {
    constructor(
        address: Address,
        private querier: () => MarketQuerier,
        client?: SigningCosmWasmClient,
        fee?: Fee,
        memo?: string,
    ) {
        super(address, client, fee, memo)
    }

    /**
     * Convert and burn the specified amount of slToken to the underlying asset
     * based on the current exchange rate and transfer them to the user.
     */
    async redeem_from_sl(burn_amount: Uint256): Promise<ExecuteResult> {
        const msg = {
            redeem_token: {
                burn_amount
            }
        }

        return this.run(msg, '60000')
    }

    /**
     * Burns slToken amount of tokens equivalent to the specified amount based on the
     * current exchange rate and transfers the specified amount of the underyling asset to the user.
     */
    async redeem_from_underlying(receive_amount: Uint256): Promise<ExecuteResult> {
        const msg = {
            redeem_underlying: {
                receive_amount
            }
        }

        return this.run(msg, '60000')
    }

    async borrow(amount: Uint256): Promise<ExecuteResult> {
        const msg = {
            borrow: {
                amount
            }
        }

        return this.run(msg, '80000')
    }

    async transfer(amount: Uint256, recipient: Address): Promise<ExecuteResult> {
        const msg = {
            transfer: {
                amount,
                recipient
            }
        }

        return this.run(msg, '80000')
    }

    /**
     * This function is automatically called before every transaction to update to
     * the latest state of the market but can also be called manually through here.
     */
    async accrue_interest(): Promise<ExecuteResult> {
        const msg = {
            accrue_interest: { }
        }

        return this.run(msg, '40000')
    }

    async deposit(amount: Uint256, underlying_asset?: Address): Promise<ExecuteResult> {
        if (!underlying_asset) {
            underlying_asset = await this.get_underlying_asset()
        }

        const snip20 = new Snip20Contract(underlying_asset, this.client)
        const fee = this.fee || create_fee('60000')

        return snip20.exec(fee, this.memo).send(this.address, amount, 'deposit')
    }

    /**
     * @param borrower - Optionally specify a borrower ID to repay someone else's debt.
     */
    async repay(amount: Uint256, borrower?: string, underlying_asset?: Address): Promise<ExecuteResult> {
        if (!underlying_asset) {
            underlying_asset = await this.get_underlying_asset()
        }

        const snip20 = new Snip20Contract(underlying_asset, this.client)
        const fee = this.fee || create_fee('90000')

        const msg = {
            repay: {
                borrower
            }
        }

        return snip20.exec(fee, this.memo).send(this.address, amount, msg)
    }

    async liquidate(
        amount: Uint256,
        borrower: string,
        collateral: Address,
        underlying_asset?: Address
    ): Promise<ExecuteResult> {
        if (!underlying_asset) {
            underlying_asset = await this.get_underlying_asset()
        }

        const snip20 = new Snip20Contract(underlying_asset, this.client)
        const fee = this.fee || create_fee('130000')

        const msg = {
            liquidate: {
                borrower,
                collateral
            }
        }

        return snip20.exec(fee, this.memo).send(this.address, amount, msg)
    }

    private async get_underlying_asset(): Promise<Address> {
        const result = await this.querier().underlying_asset()

        return result.address
    }
}

class MarketQuerier extends Querier {
    async token_info(): Promise<TokenInfo> {
        return new Snip20Contract(this.address, undefined, this.client)
            .query()
            .get_token_info()
    }

    async balance(address: Address, key: ViewingKey): Promise<Uint128> {
        return new Snip20Contract(this.address, undefined, this.client)
            .query()
            .get_balance(address, key)
    }

    async underlying_balance(auth: LendAuth): Promise<Uint128> {
        const msg = {
            balance_underlying: {
                block: (await this.client.getBlock()).header.height,
                method: await auth.create_method<MarketPermissions>(this.address, 'balance')
            }
        }

        return this.run(msg)
    }

    async state(): Promise<MarketState> {
        const msg = {
            state: {
                block: (await this.client.getBlock()).header.height
            }
        }

        return this.run(msg)
    }

    async underlying_asset(): Promise<ContractInfo> {
        const msg = {
            underlying_asset: { }
        }

        return this.run(msg)
    }

    async borrow_rate(): Promise<Decimal256> {
        const msg = {
            borrow_rate: {
                block: (await this.client.getBlock()).header.height
            }
        }

        return this.run(msg)
    }

    async supply_rate(): Promise<Decimal256> {
        const msg = {
            supply_rate: {
                block: (await this.client.getBlock()).header.height
            }
        }

        return this.run(msg)
    }

    async exchange_rate(): Promise<Decimal256> {
        const msg = {
            exchange_rate: {
                block: (await this.client.getBlock()).header.height
            }
        }

        return this.run(msg)
    }

    async account(auth: LendAuth): Promise<MarketAccount> {
        const msg = {
            account: {
                block: (await this.client.getBlock()).header.height,
                method: await auth.create_method<MarketPermissions>(this.address, 'account_info')
            }
        }

        return this.run(msg)
    }

    /**
     * Will throw if the account hasn't borrowed at least once before.
     */
    async account_id(auth: LendAuth): Promise<string> {
        const msg = {
            id: {
                method: await auth.create_method<MarketPermissions>(this.address, 'id')
            }
        }

        return this.run(msg)
    }

    /**
     * Max limit is 10.
     */
    async borrowers(pagination: Pagination): Promise<PaginatedResponse<MarketBorrower>> {
        const msg = {
            borrowers: {
                block: (await this.client.getBlock()).header.height,
                pagination
            }
        }

        return this.run(msg)
    }
}
